### PR TITLE
ASM-6091: Updated and fixed tagged vlan issue

### PR DESCRIPTION
**Pull request by prem548**
made changes on code and added test cases as per commented let me know changes if required.

### CODE REVIEW

**Comment by gavin-scott** [comment]
Generally speaking formatting still needs to be fixed.

File: spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    it "should add tagged vlans" do
+      expect(transport).to receive(:command).with("config").ordered
+      expect(transport).to receive(:command).with("interface Te1/0/29").ordered
+      expect(transport).to receive(:command).with("switchport mode general").ordered
+      expect(transport).to receive(:command).with("switchport general allowed vlan add 20 tagged").ordered
+      expect(transport).to receive(:command).with("show running-config interface Te1/0/29").ordered
+      base.update_vlans(transport,[[],[]],["Te1",0,29],"20",true)
+    end
+
```
Comment (gavin-scott) [inline spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb]: I'd like to see a test where we both remove and add tagged vlans in the same go.

File: spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      expect(transport).to receive(:command).with("show running-config interface Te1/0/29").ordered
+      base.update_vlans(transport,[[],[]],["Te1",0,29],"29",false)
+    end
+
+    it "should not add untagged vlan when it is already exist" do
+      expect(transport).to receive(:command).with("show running-config interface Te1/0/29").ordered
+      base.update_vlans(transport,[[],[29]],["Te1",0,29],"29",false)
+    end
+
+    it "should unset untagged vlans and update untagged vlans" do
```
Comment (gavin-scott) [inline spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb]: I would just call this "should set untagged vlan" since in the untagged case you are not doing any "unsetting".

File: lib/puppet_x/dell_powerconnect/model/interface/base.rb [hunk 3]
```diff
@@ -145,4 +150,84 @@ def self.register(base)
...
+    else
+      vlans_to_add = value - current_vlans
+      vlans_to_add.each do |val|
+        Puppet.debug("Adding vlans #{val}")
+        transport.command("interface #{interface_val[0]}/#{interface_val[1]}/#{interface_val[2]}")
+        transport.command("switchport mode general")
+        transport.command("switchport general pvid #{val}")
+      end
+      transport.command("show running-config interface #{interface_val[0]}/#{interface_val[1]}/#{interface_val[2]}")
+    end
```
Comment (gavin-scott) [inline lib/puppet_x/dell_powerconnect/model/interface/base.rb]: I general this method is kind of hard to follow since it's doing slightly different things for untagged case (just set the untagged vlan) and the tagged case (remove differing tagged vlans and add any missing tagged vlans). Maybe it would make more sense to just break this into separate methods for tagged and untagged in that case. I can sit with you and try to refactor it.

**Comment by gavin-scott** [comment]
@prem548 please do a couple more end-to-end tests now that we have updated this PR. Other than that, looks good.

**Comment by gavin-scott** [comment]
Also, please submit separate PR with the Windows Gemfile changes you made to run the tests on Windows.

**Comment by prem548** [comment]
finished end to end tests on power connect. It seems it was able to remove extra tagged vlans. Made a PR on updated Gemfile.

File: spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    it "should return an empty tagged and unatagged list when there are no vlans" do
+      vlan_info = PuppetSpec.load_fixture("show_interfaces_switchport/vlan_no_info.out")
+      transport.stub(:command).with("show running-config interface Te1/0/29").and_return(vlan_info)
+      expect(base.show_vlans(transport, interface_type)).to eq([[], []])
+    end
+  end
+
+  describe "#update_tagged_vlans" do
+    it "should unset extra tagged vlans" do
+      #tagged vlans for Example value="20" or value="20,23"
```
Comment (gavin-scott) [inline spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb]: @sushilrai  -- @prem548 has addressed everything I asked him to on this PR and done end-to-end testing. The goal is to remove "extra" tagged VLANs from powerconnect server-facing ports during ASM deployment. Can you give a quick once-over of the commands that are being issued in the unit tests below and make sure they look ok to you? Thanks!

Comment (sushilrai) [inline spec/unit/puppet_x/dell_powerconnect/model/interface/base_spec.rb]: Command looks good for General mode that we have been supporting till now with PowerConnect and N-Series switches.
